src/Storages/StorageDelta.cpp

```diff
@@ -0,0 +1,354 @@
+#include <Common/config.h>
+
+#if USE_AWS_S3
+
+#include <Storages/StorageDelta.h>
+#include <Common/logger_useful.h>
+
+#include <IO/ReadBufferFromS3.h>
+#include <IO/ReadSettings.h>
+#include <IO/S3Common.h>
+
+#include <Storages/StorageFactory.h>
+#include <Storages/checkAndGetLiteralArgument.h>
+
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/ListObjectsV2Request.h>
+
+#include <QueryPipeline/Pipe.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int S3_ERROR;
+    extern const int BAD_ARGUMENTS;
+}
+
+void DeltaLakeMetadata::add(const String & key, uint64_t timestamp)
+{
+    file_update_time[key] = timestamp;
+}
+
+void DeltaLakeMetadata::remove(const String & key, uint64_t /*timestamp */)
+{
+    file_update_time.erase(key);
+}
+
+std::vector<String> DeltaLakeMetadata::ListCurrentFiles() &&
+{
+    std::vector<String> keys;
+    keys.reserve(file_update_time.size());
+
+    for (auto && [k, _] : file_update_time)
+    {
+        keys.push_back(k);
+    }
+
+    return keys;
```

Comment:
```suggestion
    auto keys_view = std::views::keys(file_update_time));
    return std::vector<String>(keys_view.begin(), keys_view.end());
```

Reply:
Not working for me: "no member named 'keys' in namespace 'std::ranges::views'", ranges header is included.